File: Luca/app/src/main/java/de/culture4life/luca/ui/ViewError.java
package de.culture4life.luca.ui;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;

import javax.net.ssl.SSLPeerUnverifiedException;

import de.culture4life.luca.R;
import de.culture4life.luca.crypto.DailyKeyUnavailableException;
import de.culture4life.luca.util.ThrowableUtil;
import hu.akarnokd.rxjava3.debug.RxJavaAssemblyException;
import io.reactivex.rxjava3.core.Completable;

public class ViewError {

    @NonNull
    private String title;

    @NonNull
    private String description;

    @Nullable
    private Throwable cause;

    @Nullable
    private String resolveLabel;

    @Nullable
    private Completable resolveAction;

    private boolean removeWhenShown;

    private boolean canBeShownAsNotification;

    private boolean isExpected;

    private boolean isCancelable = true;

    public ViewError(@NonNull String title, @NonNull String description) {
        this.title = title;
        this.description = description;
        this.removeWhenShown = true;
    }

    public boolean isResolvable() {
        return resolveAction != null;
    }

    public Completable resolve() {
        return Completable.defer(() -> {
            if (isResolvable()) {
                return resolveAction;
            } else {
                return Completable.error(new IllegalStateException("Error is not resolvable"));
            }
        });
    }

    @Override
    public String toString() {
        return "ViewError{" +
                "cause=" + cause +
                ", resolveLabel='" + resolveLabel + '\'' +
                '}';
    }

    @NonNull
    public String getTitle() {
        return title;
    }

    public void setTitle(@NonNull String title) {
        this.title = title;
    }

    @NonNull
    public String getDescription() {
        return description;
    }

    public void setDescription(@NonNull String description) {
        this.description = description;
    }

    @Nullable
    public Throwable getCause() {
        return cause;
    }

    public void setCause(@Nullable Throwable cause) {
        this.cause = cause;
    }

    @Nullable
    public String getResolveLabel() {
        return resolveLabel;
    }

    public void setResolveLabel(@Nullable String resolveLabel) {
        this.resolveLabel = resolveLabel;
    }

    @Nullable
    public Completable getResolveAction() {
        return resolveAction;
    }

    public void setResolveAction(@Nullable Completable resolveAction) {
        this.resolveAction = resolveAction;
    }

    public boolean getRemoveWhenShown() {
        return removeWhenShown;
    }

    public void setRemoveWhenShown(boolean removeWhenShown) {
        this.removeWhenShown = removeWhenShown;
    }

    public boolean canBeShownAsNotification() {
        return canBeShownAsNotification;
    }

    public void setCanBeShownAsNotification(boolean canBeShownAsNotification) {
        this.canBeShownAsNotification = canBeShownAsNotification;
    }

    public boolean isExpected() {
        return isExpected;
    }

    public void setExpected(boolean expected) {
        isExpected = expected;
    }

    public boolean isCancelable() {
        return isCancelable;
    }

    public void setCancelable(boolean cancelable) {
        isCancelable = cancelable;
    }

    private static boolean isGenericException(@NonNull Throwable throwable) {
        return throwable instanceof NullPointerException;
    }

    private static String createTitle(@NonNull Throwable throwable, @NonNull Context context) {
        if (isGenericException(throwable)) {
            return context.getString(R.string.error_generic_title);
        } else if (ThrowableUtil.isNetworkError(throwable)) {
            return context.getString(R.string.error_no_internet_connection_title);
        } else if (ThrowableUtil.isCause(DailyKeyUnavailableException.class, throwable)) {
            return context.getString(R.string.error_no_daily_key_available_title);
        } else if (throwable instanceof SSLPeerUnverifiedException) {
            return context.getString(R.string.error_certificate_pinning_title);
        } else {
            String type = throwable.getClass().getSimpleName();
            return context.getString(R.string.error_specific_title, type);
        }
    }

    private static String createDescription(@NonNull Throwable throwable, @NonNull Context context) {
        if (ThrowableUtil.isNetworkError(throwable)) {
            return context.getString(R.string.error_no_internet_connection_description);
        } else if (ThrowableUtil.isCause(DailyKeyUnavailableException.class, throwable)) {
            return context.getString(R.string.error_no_daily_key_available_description);
        } else if (throwable instanceof SSLPeerUnverifiedException) {
            return context.getString(R.string.error_certificate_pinning_description);
        } else {
            String description = getMessagesFromThrowableAndCauses(throwable);
            if (description != null) {
                return context.getString(R.string.error_specific_description, description);
            } else {
                return context.getString(R.string.error_generic_description);
            }
        }
    }

    @Nullable
    public static String getMessagesFromThrowableAndCauses(@NonNull Throwable throwable) {
        if (throwable instanceof RxJavaAssemblyException) {
            // these don't have any meaningful messages
            return null;
        }
        String message = throwable.getLocalizedMessage();
        if (message == null) {
            message = throwable.getClass().getSimpleName();
        }
        if (!message.endsWith(".")) {
            message += ".";
        }
        if (throwable.getCause() != null) {
            String causeMessage = getMessagesFromThrowableAndCauses(throwable.getCause());
            if (causeMessage != null) {
                message += " " + causeMessage;
            }
        }
        return message;
    }

    public static class Builder {

        private final Context context;

        private String title;
        private String description;
        private Throwable cause;
        private String resolveLabel;
        private Completable resolveAction;
        private boolean removeWhenShown;
        private boolean canBeShownAsNotification;
        private boolean isExpected;
        private boolean isCancelable = true;

        public Builder(@NonNull Context context) {
            this.context = context;
        }

        public Builder withTitle(@NonNull String title) {
            this.title = title;
            return this;
        }

        public Builder withTitle(@StringRes int stringResource) {
            return withTitle(context.getString(stringResource));
        }

        public Builder withDescription(@NonNull String description) {
            this.description = description;
            return this;
        }

        public Builder withDescription(@StringRes int stringResource) {
            return withDescription(context.getString(stringResource));
        }

        public Builder withCause(@NonNull Throwable cause) {
            this.cause = cause;
            if (title == null) {
                title = createTitle(cause, context);
            }
            if (description == null) {
                description = createDescription(cause, context);
            }
            return this;
        }

        public Builder withResolveLabel(@NonNull String resolveLabel) {
            this.resolveLabel = resolveLabel;
            return this;
        }

        public Builder withResolveLabel(@StringRes int stringResource) {
            return withResolveLabel(context.getString(stringResource));
        }

        public Builder withResolveAction(@NonNull Completable resolveAction) {
            this.resolveAction = resolveAction;
            return this;
        }

        public Builder removeWhenShown() {
            removeWhenShown = true;
            return this;
        }

        public Builder canBeShownAsNotification() {
            canBeShownAsNotification = true;
            return this;
        }

        public Builder isExpected() {
            isExpected = true;
            return this;
        }

        public Builder setExpected(boolean isExpected) {
            this.isExpected = isExpected;
            return this;
        }

        @NonNull
        public Builder setNotCancelable() {
            isCancelable = false;
            return this;
        }

        public ViewError build() {
            if (title == null) {
                throw new IllegalStateException("No error title set");
            } else if (description == null) {
                throw new IllegalStateException("No error description set");
            } else if (resolveAction != null && resolveLabel == null) {
                throw new IllegalStateException("No resolve label set");
            }

            ViewError viewError = new ViewError(title, description);
            viewError.setCause(cause);
            viewError.setResolveAction(resolveAction);
            viewError.setResolveLabel(resolveLabel);
            viewError.setRemoveWhenShown(removeWhenShown);
            viewError.setCanBeShownAsNotification(canBeShownAsNotification);
            viewError.setExpected(isExpected);
            viewError.setCancelable(isCancelable);

            return viewError;
        }

    }

}

File: Luca/app/src/main/java/de/culture4life/luca/ui/ViewError.kt
package de.culture4life.luca.ui

import android.content.Context
import androidx.annotation.StringRes
import de.culture4life.luca.R
import de.culture4life.luca.crypto.DailyKeyUnavailableException
import de.culture4life.luca.rollout.RolloutException
import de.culture4life.luca.util.ThrowableUtil.isCause
import de.culture4life.luca.util.ThrowableUtil.isNetworkError
import de.culture4life.luca.util.getMessagesFromThrowableAndCauses
import de.culture4life.luca.util.isHttpException
import de.culture4life.luca.util.isNetworkError
import io.reactivex.rxjava3.core.Completable
import java.net.HttpURLConnection
import javax.net.ssl.SSLPeerUnverifiedException

data class ViewError(
    var title: String,
    val description: String,
    val resolveLabel: String? = null,
    val removeWhenShown: Boolean = false,
    val canBeShownAsNotification: Boolean = false,
    val isExpected: Boolean = false,
    val isCancelable: Boolean = true,
    val resolveAction: Completable? = null,
    val cause: Throwable? = null
) {
    val isResolvable: Boolean
        get() = resolveAction != null

    class Builder(private val context: Context) {

        private var title: String? = null
        private var description: String? = null
        private var cause: Throwable? = null
        private var resolveLabel: String? = null
        private var resolveAction: Completable? = null
        private var removeWhenShown = false
        private var canBeShownAsNotification = false
        private var isExpected = false
        private var isCancelable = true

        fun withTitle(title: String) = apply { this.title = title }
        fun withTitle(@StringRes stringResource: Int) = apply { withTitle(context.getString(stringResource)) }
        fun withDescription(description: String?) = apply { this.description = description }
        fun withDescription(@StringRes stringResource: Int) = apply { withDescription(context.getString(stringResource)) }
        fun withResolveLabel(resolveLabel: String) = apply { this.resolveLabel = resolveLabel }
        fun withResolveLabel(@StringRes stringResource: Int) = withResolveLabel(context.getString(stringResource))
        fun withResolveAction(resolveAction: Completable) = apply { this.resolveAction = resolveAction }
        fun removeWhenShown() = apply { removeWhenShown = true }
        fun canBeShownAsNotification() = apply { canBeShownAsNotification = true }
        fun isExpected() = apply { isExpected = true }
        fun setExpected(isExpected: Boolean) = apply { this.isExpected = isExpected }
        fun setCancelable(isCancelable: Boolean) = apply { this.isCancelable = isCancelable }
        fun withCause(cause: Throwable) = apply {
            this.cause = cause
            if (title == null) title = createTitle(cause)
            if (description == null) description = createDescription(cause)
        }

        fun build(): ViewError {
            requireNotNull(title) { "No error title set" }
            requireNotNull(description) { "No error description set" }
            check(!(resolveAction != null && resolveLabel == null)) { "No resolve label set" }
            return ViewError(
                title = title!!,
                description = description!!,
                resolveLabel = resolveLabel,
                removeWhenShown = removeWhenShown,
                canBeShownAsNotification = canBeShownAsNotification,
                isExpected = isExpected,
                isCancelable = isCancelable,
                resolveAction = resolveAction,
                cause = cause
            )
        }

        private fun createTitle(throwable: Throwable): String {
            return when {
                throwable is NullPointerException -> context.getString(R.string.error_generic_title)
                isNetworkError(throwable) -> context.getString(R.string.error_no_internet_connection_title)
                isCause(DailyKeyUnavailableException::class.java, throwable) -> context.getString(R.string.error_no_daily_key_available_title)
                throwable is SSLPeerUnverifiedException -> context.getString(R.string.error_certificate_pinning_title)
                isCause(RolloutException::class.java, throwable) -> context.getString(R.string.error_rollout_title)
                else -> {
                    val type = throwable.javaClass.simpleName
                    context.getString(R.string.error_specific_title, type)
                }
            }
        }

        private fun createDescription(throwable: Throwable): String {
            return when {
                throwable.isNetworkError() -> context.getString(R.string.error_no_internet_connection_description)
                throwable.isHttpException(
                    HttpURLConnection.HTTP_INTERNAL_ERROR,
                    HttpURLConnection.HTTP_BAD_GATEWAY,
                    HttpURLConnection.HTTP_UNAVAILABLE,
                    HttpURLConnection.HTTP_GATEWAY_TIMEOUT
                ) -> context.getString(R.string.error_http_server_error_description)
                isCause(DailyKeyUnavailableException::class.java, throwable) -> context.getString(R.string.error_no_daily_key_available_description)
                throwable is SSLPeerUnverifiedException -> context.getString(R.string.error_certificate_pinning_description)
                isCause(RolloutException::class.java, throwable) -> context.getString(R.string.error_rollout_description)
                else -> {
                    val description = throwable.getMessagesFromThrowableAndCauses()
                    if (description != null) {
                        context.getString(R.string.error_specific_description, description)
                    } else {
                        context.getString(R.string.error_generic_description)
                    }
                }
            }
        }
    }
}
